internal/safepath/join_linux.go

```diff
@@ -0,0 +1,145 @@
+package safepath
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/containerd/containerd/log"
+	"github.com/docker/docker/internal/unix_noeintr"
+	"github.com/moby/sys/mount"
+	"github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+)
+
+// Join makes sure that the concatenation of path and subpath doesn't
+// resolve to a path outside of path and returns a path to a temporary file that is
+// a bind mount to the the exact same file/directory that was validated.
+//
+// After the returned path is consumed by the caller, it is his responsibility
+// to call Close on the returned SafePath object which will unmount the
+// temporary file/directory and remove it.
+func Join(path, subpath string) (*SafePath, error) {
+	base, subpart, err := evaluatePath(path, subpath)
+	if err != nil {
+		return nil, err
+	}
+
+	fd, err := safeOpenFd(base, subpart)
+	if err != nil {
+		return nil, err
+	}
+
+	defer unix_noeintr.Close(fd)
+
+	tmpMount, err := tempMountPoint(fd)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create temporary file for safe mount")
+	}
+
+	mountSource := procFsFdPath(fd)
+	if err := mount.Mount(mountSource, tmpMount, "none", "bind"); err != nil {
+		os.Remove(tmpMount)
+		return nil, errors.Wrap(err, "failed to mount resolved path")
+	}
+
+	return &SafePath{
+		path:          tmpMount,
+		sourceBase:    base,
+		sourceSubpath: subpart,
+		cleanup:       cleanupSafePath(tmpMount),
+	}, nil
+}
+
+// safeOpenFd opens the file at filepath.Join(path, subpath) in O_PATH
+// mode and returns the file descriptor if subpath is within the subtree
+// rooted at path. It is an error if any of components of path or subpath
+// are symbolic links.
+//
+// It is a caller's responsibility to close the returned file descriptor, if no
+// error was returned.
+func safeOpenFd(path, subpath string) (int, error) {
+	// Open base volume path (_data directory).
+
+	prevFd, err := unix_noeintr.Open(path, unix.O_PATH|unix.O_DIRECTORY|unix.O_CLOEXEC|unix.O_NOFOLLOW, 0)
+	if err != nil {
+		return -1, ErrNotAccessible
```

Comment:
Seems like this could lead to misunderstood problems since we are masking the actual error.
Maybe use the new multi-error support?

Reply:
I think I'll just turn it into a rich error struct at this point that will contain the original error and the specific path information.